### ISSUE TITLE
Support Pony ORM

### ISSUE THREAD

**Issue by littlepea** [proofit404/mappers#250]
Is there a way to write a custom mapper that works with any ORM without forking the project?

I was surprised it only works with Django...

I thought mappers would just map dict keys or any class attributes to entities based on config, why does it require a specific "ORM support"?

It seems to me that if I want to have the same approach with a different ORM I should just use [Pydantic in orm_mode](https://pydantic-docs.helpmanual.io/usage/models/#orm-mode-aka-arbitrary-class-instances) and implement my own decorator instead of `mappers`, no?

**Comment by proofit404** [comment]
Hi, thank you for the question!

There are many reasons I wrote this package this way:
1. It analyzes entities against the schema. For example, it protects non-optional entity fields against nullable fields in the schema.
2. Performance. It does not fetch an ORM instance from the database. It does produce entities.
3. Nested entities support.

TLDR yes, it is necessary to support any kind of schema (orm, jsonscema, graphql) inside the project.

It requires to write two relatively short modules for schema introspection and return value instrumentation. Analysis engine and config interpretation are not coupled to Django in any way.

**Comment by littlepea** [comment]
I see, still, I'm not sure why you close this issue given that you have a bunch of similar issues currently open: #153 #1 etc...

**Comment by proofit404** [comment]
My bed. I misread Pony ORM as Peewee ORM and thought it was a question.

**Comment by littlepea** [comment]
Thanks!

It was also a question, which you answered :)

FYI here's what I've used for Pony ORM instead of mappers, for now:

```python
# Pydantic entities


class User(BaseModel):
    id: UUID
    first_name: str
    last_name: str
    username: str

    class Config:
        orm_mode = True


class Post(BaseModel):
    id: UUID
    user: User
    title: str = None
    created_at: datetime = None

    class Config:
        orm_mode = True


# ORM models


class UserDBModel(db.Entity):
    id = PrimaryKey(UUID, auto=True)
    first_name = Required(str)
    last_name = Required(str)
    username = Required(str)
    posts = Set('PostDBModel')


class PostDBModel(db.Entity):
    id = PrimaryKey(UUID, auto=True)
    user_id = Required(UserDBModel)
    created_at = Required(datetime)
    title = Optional(str)


# Data access

@db_session
@mapper(entity=Post, config={'user': 'user_id'})
def get_recent_posts(limit: int = 10) -> Iterable[Post]:
    return PostDBModel.select().order_by(desc(PostDBModel.created_at))[:limit]


# Mapper


import functools


class MapperEntityMissing(BaseException):
    pass


def mapper(function=None, *, entity=None, config=None):
    """
    Handles conversion from ORM to domain entities
    """
    if function is None:
        return functools.partial(mapper, entity=entity, config=config)

    if not entity:
        raise MapperEntityMissing

    config = config or {}

    @functools.wraps(function)
    def wrapper(*args, **kwargs):
        nonlocal entity

        entities = []
        results = function(*args, **kwargs)
        entity.update_forward_refs()
        for orm_object in results:
            for field, column in config.items():
                if hasattr(orm_object, column):
                    setattr(orm_object, field, getattr(orm_object, column))
            entities.append(entity.from_orm(orm_object))

        return entities

    return wrapper
```

Note that I don't use any more advanced features of mappers (like evaluated fields, etc).

Do you think this was a reasonable solution or it would be as easy to implement Pony support for mappers (this issue)?

**Comment by proofit404** [comment]
`mappers` would take a slightly different approach to implementation. Just because of the necessary integration with the analysis engine.

**Comment by littlepea** [comment]
Which means I could do it with mappers but it would be noticeably more effort, right?

**Comment by proofit404** [comment]
I'm not sure if it would be a hard task. For example, you could check out this module https://github.com/proofit404/mappers/blob/master/src/_mappers/sources/django/analyze.py to compare code complexity to your expectations.

**Comment by littlepea** [comment]
I think it's not super complex, the main complexity here is not in the code but in:

1. Understanding what the `schema` dict should look like to be useable my `mappers`
2. Looking deeper into the internals of the models of a specific ORM

I think if someone would document creating a new source as a tutorial based on any ORM example (like SQLAlchemy) then it would much be easier for the community to close those issues.

P.S. I do realize that I'm welcome to create such tutorial myself :D

**Comment by proofit404** [comment]
Yep, any contribution is welcome :)

**Comment by proofit404** [comment]
I would like to offer to collaborate on this issue together.

If you could set up a test project and write all reader functions used in tests, I'll take the implementation part.

You can use https://github.com/proofit404/mappers/pull/225 as an example.

What do you think?
